src/rt/cover.c

```diff
@@ -145,25 +145,68 @@ static inline void cover_toggle_check_0_1_u_z(uint8_t old, uint8_t new,
-      COVER_TGL_CB_MSG(s)                                                     \
-      for (int i = 0; i < s_size; i++) {                                      \
-         uint8_t new = ((uint8_t*)signal_value(s))[i];                        \
-         uint8_t old = ((uint8_t*)signal_last_value(s))[i];                   \
-         check_fnc(old, new, toggle_01, toggle_10);                           \
-         toggle_01 += 2;                                                      \
-         toggle_10 += 2;                                                      \
-      }                                                                       \
-      COVER_TGL_SIGNAL_DETAILS(s, s_size)                                     \
-   }                                                                          \
+// Set to match word size of common PC and optimize for int64_t comparison
+#define TOGGLE_BATCH_SIZE 8
+
+// Signal size threshold to enable optimization.
+// Signals larger than this size are considered to be less likely to have
+// many change at one callback
+#define TOGGLE_OPT_TH 128
+
+// Callback is optimized for performance
+// Check only group of 8 bytes that do have change of signal value
+#define DEFINE_COVER_TOGGLE_CB(name, check_fnc)                                     \
+   static void name(uint64_t now, rt_signal_t *s, rt_watch_t *w, void *user)        \
+   {                                                                                \
+      uint32_t s_size = s->shared.size;                                             \
+      rt_model_t *m = get_model();                                                  \
+      const int32_t tag = (uintptr_t)user;                                          \
+      COVER_TGL_CB_MSG(s)                                                           \
+      /* For small sized signals, always iterate and assume most bits change */     \
+      if (s_size <= TOGGLE_OPT_TH) {                                                \
+         int32_t *toggle_01 = get_cover_counter(m, tag);                            \
+         int32_t *toggle_10 = toggle_01 + 1;                                        \
+         for (int i = 0; i < s_size; i++) {                                         \
+            uint8_t new = ((uint8_t*)signal_value(s))[i];                           \
+            uint8_t old = ((uint8_t*)signal_last_value(s))[i];                      \
+            check_fnc(old, new, toggle_01, toggle_10);                              \
+            toggle_01 += 2;                                                         \
+            toggle_10 += 2;                                                         \
+         }                                                                          \
+         return;                                                                    \
+      }                                                                             \
```

Comment:
Why bother with this? For small sizes it falls through to the scalar loop on line 198 which does exactly the same thing. For sizes in 8..128 there's presumably some benefit to comparing eight bytes at a time, certainly more than the overhead of the extra branches.

Reply:
The idea was that in a most designs out there, there are either 32 or 64 bit SoC buses. Such buses likely change most bits of the vector during a single event (e.g. a new instruction fetched from memory). Then it is redundant to do the word compare first, and then walk each byte on compare as we well almost always need to do the walk.

Reply:
And I chose 128 with some reserve.

I think a better optimization could be based not on signal size, but based on signal being multidimensional or not. The multidimensional signals are likely to be memories where only single row changes at a given time.

Reply:
I would keep the current format, but it is up to you. I can remove the threshold completely.

Reply:
Have you actually measured any performance improvement by doing this? Here's the NEORV32 test bench run with `--cover=toggle` and the code in this PR:

```
  Time (mean ± σ):     7073.3 ms ±  53.7 ms    [User: 7054.7 ms, System: 17.6 ms]              
  Range (min … max):   7021.0 ms … 7142.0 ms    5 runs                              
```

And if I set `TOGGLE_OPT_TH` to zero:

```
  Time (mean ± σ):     7010.8 ms ±  59.8 ms    [User: 6992.7 ms, System: 15.8 ms]
  Range (min … max):   6964.1 ms … 7108.9 ms    5 runs
```

Which is actually slightly faster (although not statistically significant).

For small arrays the main cost seems to be the cache miss loading the counter before incrementing as it's indexed by the tag which is essentially random and so hard to prefetch. Once you've loaded the eight bytes of signal data it's already in L1 and the cost of loading it again a byte at a time is only around four cycles on modern CPUs which is pretty insignificant. 

On the other hand removing the `TOGGLE_OPT_TH` check increases the size and complexity of this macro by about 50% which is definitely not worth it.

Reply:
> Have you actually measured any performance improvement by doing this?

No. It was last minute change trying to avoid extra work in cases where it will likely not be needed. I should have done some benchmark.

I only measured the dummy example above when I was thinking on how to optimize the code.

> For small arrays the main cost seems to be the cache miss loading the counter before incrementing as it's indexed by the tag which is essentially random and so hard to prefetch. Once you've loaded the eight bytes of signal data it's already in L1 and the cost of loading it again a byte at a time is only around four cycles on modern CPUs which is pretty insignificant.
 
Well, that makes sense, and I did not think about it that way. Doing the 64 bit compare is one instruction, so no overhead. Loading of signal data needs to be done either way, so no savings.

> On the other hand adding the TOGGLE_OPT_TH check increases the size and complexity of this macro by about 50% which is definitely not worth it.

Yeah, it is ugly, and as it has no provable benefit I will rather remove it.